src/elliptic/curves/wrappers.rs

```diff
@@ -0,0 +1,1692 @@
+        (_o<> &Scalar<E>, Scalar<E>), (_o<> &Scalar<E>, ScalarZ<E>),
+        (r_<> &Scalar<E>, &Scalar<E>), (r_<> &Scalar<E>, &ScalarZ<E>),
+        (_o<> &ScalarZ<E>, Scalar<E>), (_o<> &ScalarZ<E>, ScalarZ<E>),
+        (r_<> &ScalarZ<E>, &Scalar<E>), (r_<> &ScalarZ<E>, &ScalarZ<E>),
+    }
+}
+
+impl<E: Curve> ops::Mul<&Scalar<E>> for Generator<E> {
+    type Output = Point<E>;
+    fn mul(self, rhs: &Scalar<E>) -> Self::Output {
+        Point::from_raw(self.as_raw().scalar_mul(&rhs.as_raw()))
+            .expect("generator multiplied by non-zero scalar is always non-zero point")
+    }
+}
+
+impl<E: Curve> ops::Mul<Scalar<E>> for Generator<E> {
+    type Output = Point<E>;
+    fn mul(self, rhs: Scalar<E>) -> Self::Output {
+        self.mul(&rhs)
+    }
+}
+
+impl<E: Curve> ops::Mul<Generator<E>> for &Scalar<E> {
+    type Output = Point<E>;
+    fn mul(self, rhs: Generator<E>) -> Self::Output {
+        rhs.mul(self)
+    }
+}
+
+impl<E: Curve> ops::Mul<Generator<E>> for Scalar<E> {
+    type Output = Point<E>;
+    fn mul(self, rhs: Generator<E>) -> Self::Output {
+        rhs.mul(self)
+    }
+}
+
+impl<E: Curve> ops::Neg for Scalar<E> {
+    type Output = Scalar<E>;
+
+    fn neg(self) -> Self::Output {
+        Scalar::from_raw(self.as_raw().neg()).expect("neg must not produce zero point")
+    }
+}
+
+impl<E: Curve> ops::Neg for &Scalar<E> {
+    type Output = Scalar<E>;
+
+    fn neg(self) -> Self::Output {
+        Scalar::from_raw(self.as_raw().neg()).expect("neg must not produce zero point")
+    }
+}
+
+impl<E: Curve> ops::Neg for ScalarZ<E> {
+    type Output = ScalarZ<E>;
+
+    fn neg(self) -> Self::Output {
+        ScalarZ::from_raw(self.as_raw().neg())
+    }
+}
+
+impl<E: Curve> ops::Neg for &ScalarZ<E> {
+    type Output = ScalarZ<E>;
+
+    fn neg(self) -> Self::Output {
+        ScalarZ::from_raw(self.as_raw().neg())
+    }
+}
+
+impl<E: Curve> ops::Neg for Point<E> {
+    type Output = Point<E>;
+
+    fn neg(self) -> Self::Output {
+        Point::from_raw(self.as_raw().neg_point()).expect("neg must not produce zero point")
+    }
+}
+
+impl<E: Curve> ops::Neg for &Point<E> {
+    type Output = Point<E>;
+
+    fn neg(self) -> Self::Output {
+        Point::from_raw(self.as_raw().neg_point()).expect("neg must not produce zero point")
+    }
+}
+
+impl<'p, E: Curve> ops::Neg for PointRef<'p, E> {
+    type Output = Point<E>;
+
+    fn neg(self) -> Self::Output {
+        Point::from_raw(self.as_raw().neg_point()).expect("neg must not produce zero point")
+    }
+}
+
+impl<E: Curve> ops::Neg for PointZ<E> {
+    type Output = PointZ<E>;
+
+    fn neg(self) -> Self::Output {
+        PointZ::from_raw(self.as_raw().neg_point())
+    }
+}
+
+impl<E: Curve> ops::Neg for &PointZ<E> {
+    type Output = PointZ<E>;
+
+    fn neg(self) -> Self::Output {
+        PointZ::from_raw(self.as_raw().neg_point())
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(bound = "")]
+struct PointFormat<E: Curve> {
```

Comment:
Why is this needed?
can't we impl Serialize/Deserialize for `Point` directly?

Reply:
PointFormat is used for both Point, PointZ serialization. Also I like serde macros doing my work :) From my experience, implementing (De)Serialize is difficult even for small structures